File: tools/pe-parser/src/bin/sig-uniqueness-checker.rs
use clap::Parser;
use pe_parser::{
    format_byte_pattern, load_binary_data, parse_byte_pattern, parse_pe_bytes, scan_pattern,
    section_data,
};
use serde::Serialize;
use std::fs;
use std::path::PathBuf;

#[derive(Parser)]
#[command(about = "Check whether a byte signature uniquely identifies one instruction")]
struct Args {
    binary: PathBuf,
    #[arg(long)]
    sig: Option<String>,
    #[arg(long = "sig-file")]
    sig_file: Option<PathBuf>,
    #[arg(long)]
    sections: Option<String>,
    #[arg(long = "near-misses", default_value_t = 0)]
    near_misses: usize,
    #[arg(long)]
    json: bool,
}

#[derive(Serialize, Clone)]
struct Hit {
    section: String,
    file_offset: u32,
    rva: u32,
    context: String,
}

#[derive(Serialize)]
struct Probe {
    wildcarded: usize,
    matches: usize,
}

#[derive(Serialize)]
struct NearMiss {
    section: String,
    file_offset: u32,
    rva: u32,
    distance: usize,
    diff_indices: Vec<usize>,
}

#[derive(Serialize)]
struct ResultRow {
    name: String,
    pattern: String,
    length: usize,
    matches: usize,
    hits: Vec<Hit>,
    verdict: String,
    hint: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    margin: Option<usize>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    probe: Vec<Probe>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    near_misses: Vec<NearMiss>,
}

fn load_sigs(args: &Args) -> Result<Vec<(String, Vec<Option<u8>>)>, String> {
    if let Some(sig) = &args.sig {
        return Ok(vec![("sig".to_string(), parse_byte_pattern(sig)?)]);
    }
    let path = args
        .sig_file
        .as_ref()
        .ok_or_else(|| "one of --sig or --sig-file is required".to_string())?;
    let text = fs::read_to_string(path).map_err(|e| format!("{e}"))?;
    let mut out = Vec::new();
    for raw in text.lines() {
        let line = raw.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let (name, sig) = line
            .split_once('=')
            .ok_or_else(|| format!("expected name=sig: {line}"))?;
        out.push((name.trim().to_string(), parse_byte_pattern(sig.trim())?));
    }
    if out.is_empty() {
        return Err("sig file has no entries".to_string());
    }
    Ok(out)
}

fn context_hex(buf: &[u8], start: usize) -> String {
    buf[start..buf.len().min(start + 16)]
        .iter()
        .map(|b| format!("{b:02x}"))
        .collect::<Vec<_>>()
        .join(" ")
}

fn hits_for(data: &[u8], sections: &[pe_parser::Section], pattern: &[Option<u8>]) -> Vec<Hit> {
    let mut hits = Vec::new();
    for sec in sections {
        let buf = section_data(data, sec);
        for idx in scan_pattern(buf, pattern) {
            hits.push(Hit {
                section: sec.name.clone(),
                file_offset: sec.raddr + idx as u32,
                rva: sec.vaddr + idx as u32,
                context: context_hex(buf, idx),
            });
        }
    }
    hits
}

fn count_for(data: &[u8], sections: &[pe_parser::Section], pattern: &[Option<u8>]) -> usize {
    sections
        .iter()
        .map(|sec| scan_pattern(section_data(data, sec), pattern).len())
        .sum()
}

fn uniqueness_margin(
    data: &[u8],
    sections: &[pe_parser::Section],
    pattern: &[Option<u8>],
) -> usize {
    let mut shortest = pattern.len();
    for len in (1..pattern.len()).rev() {
        if count_for(data, sections, &pattern[..len]) == 1 {
            shortest = len;
        } else {
            break;
        }
    }
    pattern.len() - shortest
}

fn trailing_probe(
    data: &[u8],
    sections: &[pe_parser::Section],
    pattern: &[Option<u8>],
) -> Vec<Probe> {
    [1usize, 2, 4]
        .iter()
        .filter_map(|k| {
            if *k >= pattern.len() {
                return None;
            }
            let mut probe = pattern.to_vec();
            for idx in probe.len() - *k..probe.len() {
                probe[idx] = None;
            }
            Some(Probe {
                wildcarded: *k,
                matches: count_for(data, sections, &probe),
            })
        })
        .collect()
}

fn near_misses(
    data: &[u8],
    sections: &[pe_parser::Section],
    pattern: &[Option<u8>],
    max_dist: usize,
) -> Vec<NearMiss> {
    if max_dist == 0 {
        return Vec::new();
    }
    let fixed: Vec<(usize, u8)> = pattern
        .iter()
        .enumerate()
        .filter_map(|(idx, b)| b.map(|v| (idx, v)))
        .collect();
    let mut out = Vec::new();
    for sec in sections {
        let buf = section_data(data, sec);
        if pattern.len() > buf.len() {
            continue;
        }
        for start in 0..=buf.len() - pattern.len() {
            let mut diff = Vec::new();
            for (idx, byte) in &fixed {
                if buf[start + idx] != *byte {
                    diff.push(*idx);
                    if diff.len() > max_dist {
                        break;
                    }
                }
            }
            if !diff.is_empty() && diff.len() <= max_dist {
                out.push(NearMiss {
                    section: sec.name.clone(),
                    file_offset: sec.raddr + start as u32,
                    rva: sec.vaddr + start as u32,
                    distance: diff.len(),
                    diff_indices: diff,
                });
                if out.len() >= 20 {
                    return out;
                }
            }
        }
    }
    out
}

fn analyze(
    data: &[u8],
    sections: &[pe_parser::Section],
    name: String,
    pattern: Vec<Option<u8>>,
    near: usize,
) -> ResultRow {
    let hits = hits_for(data, sections, &pattern);
    let matches = hits.len();
    let mut margin = None;
    let mut probe = Vec::new();
    let (verdict, hint) = if matches == 1 {
        let m = uniqueness_margin(data, sections, &pattern);
        margin = Some(m);
        if m == 0 {
            (
                "BRITTLE".to_string(),
                "Every byte is load-bearing; one patched byte makes it stale. Append stable trailing context for a safety margin.".to_string(),
            )
        } else if m >= 8 {
            let keep = pattern.len() - m;
            (
                "UNIQUE".to_string(),
                format!(
                    "Overspecified by {m} bytes; could shorten to \"{}\" and stay unique.",
                    format_byte_pattern(&pattern[..keep])
                ),
            )
        } else {
            (
                "UNIQUE".to_string(),
                format!("Unique with a {m}-byte margin."),
            )
        }
    } else if matches == 0 {
        probe = trailing_probe(data, sections, &pattern);
        let hit = probe.iter().find(|p| p.matches >= 1);
        let hint = match hit {
            Some(p) => format!(
                "Wildcarding the last {} byte(s) yields {} match(es) - a trailing byte changed; re-dump there.",
                p.wildcarded, p.matches
            ),
            None => "No trailing variant matches; the region moved or was removed. Re-generate.".to_string(),
        };
        ("STALE".to_string(), hint)
    } else {
        (
            "AMBIGUOUS".to_string(),
            format!("{matches} matches; lengthen the sig with following bytes to disambiguate."),
        )
    };

    ResultRow {
        name,
        pattern: format_byte_pattern(&pattern),
        length: pattern.len(),
        matches,
        hits,
        verdict,
        hint,
        margin,
        probe,
        near_misses: near_misses(data, sections, &pattern, near),
    }
}

fn print_result(row: &ResultRow) {
    println!("SIG {} = {}  ({} bytes)", row.name, row.pattern, row.length);
    println!("  Verdict: {}   Matches: {}", row.verdict, row.matches);
    for hit in &row.hits {
        println!(
            "    {:<8} off=0x{:X}  rva=0x{:X}  ctx: {}",
            hit.section, hit.file_offset, hit.rva, hit.context
        );
    }
    if let Some(margin) = row.margin {
        println!("  Margin: {margin} trailing byte(s) removable while still unique");
    }
    for probe in &row.probe {
        println!(
            "    last {} wildcarded -> {} match(es)",
            probe.wildcarded, probe.matches
        );
    }
    for nm in &row.near_misses {
        let cols = nm
            .diff_indices
            .iter()
            .map(|i| i.to_string())
            .collect::<Vec<_>>()
            .join(",");
        println!(
            "    near rva=0x{:X}  dist={}  differs at byte idx [{}] -> wildcard these to widen",
            nm.rva, nm.distance, cols
        );
    }
    println!("  Hint: {}\n", row.hint);
}

fn main() {
    let args = Args::parse();
    if args.sig.is_none() == args.sig_file.is_none() {
        eprintln!("Error: exactly one of --sig or --sig-file is required");
        std::process::exit(2);
    }
    let data = match load_binary_data(&args.binary) {
        Ok(data) => data,
        Err(e) => {
            eprintln!("Error: {e}");
            std::process::exit(1);
        }
    };
    let parsed = match parse_pe_bytes(&data) {
        Ok(parsed) => parsed,
        Err(e) => {
            eprintln!("Error: {e}");
            std::process::exit(1);
        }
    };
    let want: Option<Vec<String>> = args
        .sections
        .as_ref()
        .map(|s| s.split(',').map(|x| x.trim().to_string()).collect());
    let sections: Vec<_> = parsed
        .sections
        .iter()
        .filter(|sec| match &want {
            Some(names) => names.iter().any(|n| n == &sec.name),
            None => sec.exec,
        })
        .cloned()
        .collect();
    if sections.is_empty() {
        eprintln!("Error: no matching sections to scan");
        std::process::exit(1);
    }
    let sigs = match load_sigs(&args) {
        Ok(sigs) => sigs,
        Err(e) => {
            eprintln!("Error: {e}");
            std::process::exit(1);
        }
    };
    let rows: Vec<_> = sigs
        .into_iter()
        .map(|(name, pattern)| analyze(&data, &sections, name, pattern, args.near_misses))
        .collect();
    if args.json {
        println!("{}", serde_json::to_string_pretty(&rows).unwrap());
        return;
    }
    println!(
        "File: {}  Sections scanned: {}\n",
        args.binary.display(),
        sections
            .iter()
            .map(|s| s.name.as_str())
            .collect::<Vec<_>>()
            .join(", ")
    );
    for row in &rows {
        print_result(row);
    }
}

File: tools/sig-uniqueness-checker.py
#!/usr/bin/env python3
from __future__ import annotations
"""Check whether a byte signature uniquely identifies one instruction in a PE.

Scans the executable sections of a PE binary for a candidate pattern and reports
whether it matches 0 (stale), 1 (good), or N (ambiguous) locations. For unique
sigs it measures the uniqueness margin — how many trailing bytes are redundant;
for stale sigs it probes trailing-wildcard variants to locate the changed byte;
for any sig it can list near-misses to show which bytes to wildcard to widen.

Usage:
    python3 sig-uniqueness-checker.py <binary> --sig "48 8B ?? ?? ?? ??"
    python3 sig-uniqueness-checker.py <binary> --sig-file sigs.txt        # name=sig per line
    python3 sig-uniqueness-checker.py <binary> --sig "..." --sections .text,.rdata
    python3 sig-uniqueness-checker.py <binary> --sig "..." --near-misses 1
    python3 sig-uniqueness-checker.py <binary> --sig "..." --json
"""
import sys
import struct
import os
import json
import argparse
import subprocess



# ── PE parser imports ─────────────────────────────────────────────────────────
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from tools.lib.pe_parse import parse_pe

def get_sections(data: bytes) -> list[dict]:
    try:
        pe = parse_pe(data)
    except SystemExit:
        return []
    sections = []
    for s in pe['sections']:
        sections.append({
            'name': s['name'],
            'raddr': s['raddr'],
            'rsize': s['rsize'],
            'vaddr': s['vaddr'],
            'exec': s['exec'],
        })
    return sections


def parse_pattern(text):
    """Parse '48 8B ?? ?? ..' into a list of ints (fixed) and None (wildcard)."""
    out = []
    for tok in text.replace(',', ' ').split():
        out.append(None if tok in ('?', '??') else int(tok, 16))  # int() raises on junk
    if not out:
        raise ValueError('empty pattern')
    return out


def format_pattern(pattern):
    return ' '.join('??' if b is None else f'{b:02X}' for b in pattern)


def scan(buf, pattern):
    """Yield start indices in buf where pattern (None = wildcard) matches."""
    n, m = len(buf), len(pattern)
    if m > n:
        return
    anchor = next((i for i, b in enumerate(pattern) if b is not None), None)
    if anchor is None:                       # all-wildcard pattern matches everywhere
        yield from range(n - m + 1)
        return
    fixed = [(i, b) for i, b in enumerate(pattern) if b is not None]
    needle = bytes([pattern[anchor]])
    # Anchor on the first fixed byte and skip with bytes.find — avoids scanning
    # every offset when the leading byte is rare.
    pos = buf.find(needle, anchor)
    while pos != -1 and pos - anchor + m <= n:
        start = pos - anchor
        if all(buf[start + i] == b for i, b in fixed):
            yield start
        pos = buf.find(needle, pos + 1)


def build_regions(data, sections):
    return [(s, data[s['raddr']:s['raddr'] + s['rsize']]) for s in sections]


def hits_for(regions, pattern):
    hits = []
    for sec, buf in regions:
        for idx in scan(buf, pattern):
            hits.append({'section': sec['name'],
                         'file_offset': sec['raddr'] + idx,
                         'rva': sec['vaddr'] + idx,
                         'context': buf[idx:idx + 16].hex(' ')})
    return hits


def count_for(regions, pattern):
    return sum(1 for _, buf in regions for _ in scan(buf, pattern))


def uniqueness_margin(regions, pattern):
    """Trailing bytes removable while the prefix still matches exactly once."""
    full = len(pattern)
    shortest = full
    for length in range(full - 1, 0, -1):
        # A prefix of a unique pattern can only match the same-or-more places,
        # so once a prefix is no longer unique, shorter ones never are either.
        if count_for(regions, pattern[:length]) == 1:
            shortest = length
        else:
            break
    return full - shortest


def trailing_probe(regions, pattern):
    """For a stale sig, wildcard the last 1/2/4 bytes to find the changed tail."""
    out = []
    for k in (1, 2, 4):
        if k >= len(pattern):
            continue
        probe = pattern[:-k] + [None] * k
        out.append({'wildcarded': k, 'matches': count_for(regions, probe)})
    return out


def near_misses(regions, pattern, max_dist, cap=20):
    """Positions matching all but <= max_dist of the fixed bytes."""
    fixed = [(i, b) for i, b in enumerate(pattern) if b is not None]
    m = len(pattern)
    found = []
    for sec, buf in regions:
        # ponytail: naive O(n*m) sweep, only runs when --near-misses is set.
        for start in range(len(buf) - m + 1):
            d, diff = 0, []
            for i, b in fixed:
                if buf[start + i] != b:
                    d += 1
                    diff.append(i)
                    if d > max_dist:
                        break
            if 1 <= d <= max_dist:
                found.append({'section': sec['name'],
                              'file_offset': sec['raddr'] + start,
                              'rva': sec['vaddr'] + start,
                              'distance': d, 'diff_indices': diff})
                if len(found) >= cap:
                    return found
    return found


def analyze(regions, name, pattern, near):
    hits = hits_for(regions, pattern)
    n = len(hits)
    res = {'name': name, 'pattern': format_pattern(pattern),
           'length': len(pattern), 'matches': n, 'hits': hits}
    if n == 1:
        margin = uniqueness_margin(regions, pattern)
        res['margin'] = margin
        if margin == 0:
            res['verdict'] = 'BRITTLE'
            res['hint'] = ('Every byte is load-bearing; one patched byte makes it stale. '
                           'Append stable trailing context for a safety margin.')
        elif margin >= 8:
            keep = len(pattern) - margin
            res['verdict'] = 'UNIQUE'
            res['hint'] = (f'Overspecified by {margin} bytes; could shorten to '
                           f'"{format_pattern(pattern[:keep])}" and stay unique.')
        else:
            res['verdict'] = 'UNIQUE'
            res['hint'] = f'Unique with a {margin}-byte margin.'
    elif n == 0:
        res['verdict'] = 'STALE'
        res['probe'] = trailing_probe(regions, pattern)
        hit = next((p for p in res['probe'] if p['matches'] >= 1), None)
        if hit:
            res['hint'] = (f'Wildcarding the last {hit["wildcarded"]} byte(s) yields '
                           f'{hit["matches"]} match(es) — a trailing byte changed; re-dump there.')
        else:
            res['hint'] = 'No trailing variant matches; the region moved or was removed. Re-generate.'
    else:
        res['verdict'] = 'AMBIGUOUS'
        res['hint'] = f'{n} matches; lengthen the sig with following bytes to disambiguate.'
    if near > 0:
        res['near_misses'] = near_misses(regions, pattern, near)
    return res


def print_result(r):
    print(f"SIG {r['name']} = {r['pattern']}  ({r['length']} bytes)")
    print(f"  Verdict: {r['verdict']}   Matches: {r['matches']}")
    for h in r['hits']:
        print(f"    {h['section']:<8} off=0x{h['file_offset']:X}  rva=0x{h['rva']:X}  ctx: {h['context']}")
    if 'margin' in r:
        print(f"  Margin: {r['margin']} trailing byte(s) removable while still unique")
    for p in r.get('probe', []):
        print(f"    last {p['wildcarded']} wildcarded -> {p['matches']} match(es)")
    for nm in r.get('near_misses', []):
        cols = ','.join(str(i) for i in nm['diff_indices'])
        print(f"    near rva=0x{nm['rva']:X}  dist={nm['distance']}  "
              f"differs at byte idx [{cols}] -> wildcard these to widen")
    print(f"  Hint: {r['hint']}\n")


def load_sigs(args):
    if args.sig:
        return [('sig', parse_pattern(args.sig))]
    out = []
    with open(args.sig_file) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            name, sep, sig = line.partition('=')
            if not sep or not sig.strip():
                raise ValueError(f'expected name=sig: {line}')
            out.append((name.strip(), parse_pattern(sig.strip())))
    if not out:
        raise ValueError('sig file has no entries')
    return out


def select(sections, names):
    if names:
        want = set(names)
        return [s for s in sections if s['name'] in want]
    return [s for s in sections if s['exec']]


def main():
    # Attempt to proxy to Rust binary if compiled
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    bin_name = 'sig-uniqueness-checker.exe' if os.name == 'nt' else 'sig-uniqueness-checker'
    binary_path = os.path.join(base_dir, 'bin', bin_name)

    if os.path.exists(binary_path):
        try:
            res = subprocess.run([binary_path] + sys.argv[1:])
            sys.exit(res.returncode)
        except Exception:
            pass

    p = argparse.ArgumentParser(description='Check PE signature uniqueness')
    p.add_argument('binary', help='PE binary to scan')
    g = p.add_mutually_exclusive_group(required=True)
    g.add_argument('--sig', help='single pattern, e.g. "48 8B ?? ?? ?? ??"')
    g.add_argument('--sig-file', help='file with one name=sig per line')
    p.add_argument('--sections', help='comma-separated section names (default: all executable)')
    p.add_argument('--near-misses', type=int, default=0, metavar='N',
                   help='also report matches within Hamming distance N on fixed bytes')
    p.add_argument('--json', action='store_true', help='machine-readable output')
    args = p.parse_args()

    if not os.path.isfile(args.binary):
        print(f"Error: {args.binary} not found", file=sys.stderr)
        sys.exit(1)
    with open(args.binary, 'rb') as f:
        data = f.read()
    sections = get_sections(data)
    if not sections:
        print("Error: not a PE file", file=sys.stderr)
        sys.exit(1)

    names = [n.strip() for n in args.sections.split(',')] if args.sections else None
    selected = select(sections, names)
    if not selected:
        print("Error: no matching sections to scan", file=sys.stderr)
        sys.exit(1)
    regions = build_regions(data, selected)

    try:
        sigs = load_sigs(args)
    except (ValueError, OSError) as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    results = [analyze(regions, name, pat, args.near_misses) for name, pat in sigs]

    if args.json:
        print(json.dumps(results, indent=2))
        return
    print(f"File: {args.binary}  Sections scanned: {', '.join(s['name'] for s in selected)}\n")
    for r in results:
        print_result(r)


if __name__ == '__main__':
    main()
